tot/models.py

```diff
@@ -225,3 +231,302 @@ def maybe_drop_added_dates(self, predicted, df):
+                dataframe containing test data enlarged with K values.
+        """
+        if self.K > 0:
+            df_train, _, _, _, _ = df_utils.prep_or_copy_df(df_train)
+            (
+                df_test,
+                received_ID_col_test,
+                received_single_time_series_test,
+                received_dict_test,
+                _,
+            ) = df_utils.prep_or_copy_df(df_test)
+            df_test_new = pd.DataFrame()
+            for df_name, df_test_i in df_test.groupby("ID"):
+                df_train_i = df_train[df_train["ID"] == df_name].copy(deep=True)
+                df_test_i = pd.concat([df_train_i.tail(self.K), df_test_i], ignore_index=True)
+                df_test_new = pd.concat((df_test_new, df_test_i), ignore_index=True)
+            df_test = df_utils.return_df_in_original_format(
+                df_test_new, received_ID_col_test, received_single_time_series_test, received_dict_test
+            )
+        return df_test
+
+    def maybe_drop_first_forecasts(self, predicted, df):
+        """
+        Removes first K values from predicted and df that have been previously added.
+
+        Parameters
+        ----------
+            predicted: pd.DataFrame
+                dataframe containing predicted data
+            df: pd.DataFrame
+                dataframe containing initial data
+
+        Returns
+        -------
+            pd.DataFrame
+                dataframe containing predicted data reduced by the first K values.
+            pd.DataFrame
+                dataframe containing initial data reduced by the first K values.
+        """
+        if self.K > 0:
+            (
+                predicted,
+                received_ID_col_pred,
+                received_single_time_series_pred,
+                received_dict_test_pred,
+                _,
+            ) = df_utils.prep_or_copy_df(predicted)
+            df, received_ID_col_df, received_single_time_series_df, received_dict_test_df, _ = df_utils.prep_or_copy_df(
+                df
+            )
+            predicted_new = pd.DataFrame()
+            df_new = pd.DataFrame()
+            for df_name, df_i in df.groupby("ID"):
+                predicted_i = predicted[predicted["ID"] == df_name].copy(deep=True)
+                predicted_i = predicted_i[self.K :]
+                df_i = df_i[self.K :]
+                df_new = pd.concat((df_new, df_i), ignore_index=True)
+                predicted_new = pd.concat((predicted_new, predicted_i), ignore_index=True)
+            df = df_utils.return_df_in_original_format(
+                df_new, received_ID_col_df, received_single_time_series_df, received_dict_test_df
+            )
+            predicted = df_utils.return_df_in_original_format(
+                predicted_new, received_ID_col_pred, received_single_time_series_pred, received_dict_test_pred
+            )
+        return predicted, df
+
+    def maybe_drop_added_dates(self, predicted, df):
```

Comment:
I don't think we add any dates in this model?
so we could maybe not do anything here?